storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc

```diff
@@ -960,136 +1084,187 @@ bool xbcloud_delete(Object_store *store, const std::string &container,
+      return false;
     }
-    if (error) break;
-    for (auto it = chunks.begin(); it != chunks.end();) {
-      if (error) break;
-      if (!download_state.start_chunk(it->first)) {
-        ++it;
-        continue;
-      }
-      std::stringstream s_object_name;
-      s_object_name << it->first << "." << std::setw(20) << std::setfill('0')
-                    << it->second.next;
-      std::string object_name = s_object_name.str();
-      msg_ts("%s: Downloading %s.\n", my_progname, object_name.c_str());
-      store->async_download_object(
-          container, object_name, &h,
-          std::bind(
-              [](bool success, const Http_buffer &contents, std::string name,
-                 std::string basename, struct download_state *download_state,
-                 bool *error) {
-                if (!success) {
-                  *error = true;
-                  msg_ts("%s: Download failed. Cannot download %s.\n",
-                         my_progname, name.c_str());
-                  download_state->complete_chunk(basename);
-                  return;
-                }
-                msg_ts("%s: Download successfull %s, size %zu\n", my_progname,
-                       name.c_str(), contents.size());
-                std::cout.write(&contents[0], contents.size());
-                std::cout.flush();
-                download_state->complete_chunk(basename);
-                if (std::cout.fail()) {
-                  msg_ts(
-                      "%s: Download failed. Cannot write to the standard "
-                      "output.\n",
-                      my_progname);
-                  *error = true;
-                }
-              },
-              std::placeholders::_1, std::placeholders::_2, object_name,
-              it->first, &download_state, &error));
-      if (++it->second.next > it->second.last) {
-        it = chunks.erase(it);
-      } else {
-        ++it;
-      }
+    for (uint i = 0; i < opt_threads; i++) {
+      std::string path = "thread_" + std::to_string(i);
+      fn_format(fullpath, path.c_str(), opt_stream_dir, "",
+                MYF(MY_RELATIVE_PATH));
+      mkfifo(fullpath, 0666);
     }
+    msg_ts(
+        "Created %ld Named Pipes(FIFO). Waiting up to 10 seconds for xbstream "
+        "to "
+        "open "
+        "the files for reading.\n",
+        opt_threads);
+  }
+  /* Create data copying threads */
+  data_threads = (download_thread_ctxt_t *)my_malloc(
+      PSI_NOT_INSTRUMENTED, sizeof(download_thread_ctxt_t) * (opt_threads + 1),
+      MYF(MY_FAE));
+  threads = new pthread_t[opt_threads + 1];
+  for (uint i = 0; i < (uint)opt_threads; i++) {
+    data_threads[i].num = i;
+    data_threads[i].has_errors = &has_errors;
+    data_threads[i].file_list = file_list;
+    data_threads[i].store = store;
+    data_threads[i].container = &container;
+    pthread_create(&threads[i], nullptr, &download_func, data_threads + i);
```

Comment:
handle error if pthread_create fails.

Reply:
Done. `has_errors` will signal already started threads to cleanup and exit.